src/evict/evict_lru.c

```diff
@@ -272,11 +272,24 @@ __wt_evict_thread_run(WT_SESSION_IMPL *session, WT_THREAD *thread)
     WT_CACHE *cache;
     WT_CONNECTION_IMPL *conn;
     WT_DECL_RET;
+    uint32_t session_flags;
     bool did_work, was_intr;
+    bool is_owner;
 
     conn = S2C(session);
     cache = conn->cache;
 
+    /*
+     * Cache a history store cursor to avoid deadlock: if an eviction thread thread marks a file
+     * busy and then opens a different file (in this case, the HS file), it can deadlock with a
+     * thread waiting for the first file to drain from the eviction queue. See WT-5946 for details.
+     */
+    if (!F_ISSET(conn, WT_CONN_IN_MEMORY)) {
+        session_flags = 0; /* [-Werror=maybe-uninitialized] */
+        WT_RET(__wt_hs_cursor(session, &session_flags, &is_owner));
+        WT_RET(__wt_hs_cursor_close(session, session_flags, is_owner));
```

Comment:
We already open/close the HS cursor in __wt_evict(), which gets called through __evict_lru_pages(), so for the eviction workers we might end up caching the HS dhandle twice.

Reply:
@sulabhM, I'm sorry, but I'm missing something. Why doesn't the open-cursor call find the already cached HS dhandle? Why would it cache it twice?

Reply:
I meant we will be doing open/close twice, I think it is okay, since we will be doing that every time we evict anyways. Cursor caching should reduce the overhead.